src/cmd/ksh93/tests/vi.exp

```diff
@@ -0,0 +1,670 @@
+expect "\r\nbuh bye\r\n" {
+    puts "echo 2 works"
+}
+expect_prompt
+
+# ==========
+# [count]k  Fetch previous command.  Each time k is entered the previous command back in time is accessed.
+control_mode
+send "k"
+redraw_prompt
+expect -ex ": echo buh bye" {
+    puts "k fetches previous history item"
+}
+
+control_mode
+send "k"
+redraw_prompt
+expect -ex ": echo hello"
+
+# ==========
+# [count]j  Fetch next command.  Each time j is entered the next command forward in time is accessed.
+control_mode
+send "j"
+redraw_prompt
+expect -ex ": echo buh bye" {
+    puts "j fetches next history item"
+}
+
+# [count]-  Equivalent to k.
+
+# [count][A If cursor is at the end of the line it is equivalent to / with string^set to the contents of the current line.  Otherwise, it is equivalent to k.
+# [count]+  Equivalent to j.
+# [count][B Equivalent to j.
+# [count]G  The command number count is fetched.  The default is the least recent history command.
+
+# /string   Search backward through history for a previous command containing string.  String is terminated by a `RETURN' or `NEW LINE'.  If string is preceded by a ^, the matched line must begin with string.  If string     is null, the previous string will be used.
+# ?string   Same as / except that search will be in the forward direction.
+
+# n         Search for next match of the last pattern to / or ?  commands.
+
+# N         Search for next match of the last pattern to / or ?, but in reverse direction.
+
+# ==========
+# Take cursor to beginning of line with ^
+log_test_entry
+control_mode
+send "^"
+insert_mode
+send "x"
+redraw_prompt
+expect -ex "xecho buh bye" {
+    puts "^ takes cursor to beginning of line"
+}
+
+# ==========
+# Take cursor to the end of line with $
+log_test_entry
+control_mode
+send "$"
+insert_mode
+send "Y"
+redraw_prompt
+expect -ex "xecho buh byYe" {
+    puts "$ takes cursor to end of line"
+}
+
+# ==========
+# Move cursor backward one character with h
+log_test_entry
+control_mode
+# Move to end of line
+send "$"
+send "h"
+insert_mode
+send "X"
+redraw_prompt
+expect -ex "xecho buh byXYe" {
+    puts "h takes cursor one character backward"
+}
+
+# ==========
+# Move cursor forward one character with l
+log_test_entry
+control_mode
+send "^"
+send "l"
+insert_mode
+send "z"
+redraw_prompt
+expect -ex "xzecho buh byXYe" {
+    puts "l takes cursor one character forward"
+}
+
+# ==========
+# Move cursor forward one word with W
+log_test_entry
+control_mode
+send "^"
+send "W"
+insert_mode
+send "1"
+redraw_prompt
+expect -ex "xzecho 1buh byXYe" {
+    puts "W takes cursor one word forward"
+}
+
+# ==========
+# Move cursor forward one alpha-numeric word with w
+log_test_entry
+control_mode
+send "^"
+insert_mode
+send "1"
+control_mode
+send "w"
+insert_mode
+send "2"
+redraw_prompt
+expect -ex "1xzecho 21buh byXYe" {
+    puts "W takes cursor one alpha-numeric word forward"
+}
+
+exec mkdir foo1 foo2 foo3
+
+# [count]e  Cursor to end of word
+clear_prompt
```

Comment:
Every test that ends without leaving a fresh prompt should call `clear_prompt`. In other words, a test like this one should not begin with `clear_prompt`. It should be able to assume the shell has issued a prompt and is waiting for input. If `clear_prompt` is needed it should be tied to the end of the previous test. Also, you need to call `expect_prompt` after, or as part of, `clear_prompt` to ensure we're correctly keeping track of the prompts.

Lastly, each independent test like this one should begin with `log_test_entry` to inject markers in the *interactive.tmp.log* that help when debugging failures.

Reply:
Calling `expect_prompt` inside `clear_prompt` starts causing tests to fail. I am still not sure why.